File: 525. Contiguous Array.java
class Solution {
    public int findMaxLength(int[] nums) {
        int sum = 0;
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) sum--;
            else sum++;

            if (sum == 0) {
                answer = Math.max(answer, i+1);
            }

            if (map.containsKey(sum)) {
                answer = Math.max(i - map.get(sum), answer);
            } else {
                map.put(sum, i);
            }
        }
        return answer;
    }
}

File: 525. Contiguous Array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        n = len(nums)
        prefix = {0: -1}
        sum, answer = 0, 0
        for i in range(0, n):
            sum += 1 if nums[i] == 1 else -1
            if sum in prefix:
                answer = max(answer, i - prefix[sum])
            else:
                prefix[sum] = i
        return answer
